diff --git a/include/asm-generic/dma-mapping-common.h b/include/asm-generic/dma-mapping-common.h
index 3bee2e334ad4..5ca8f8bafb4e 100644
--- a/include/asm-generic/dma-mapping-common.h
+++ b/include/asm-generic/dma-mapping-common.h
@@ -6,6 +6,7 @@
 #include <linux/scatterlist.h>
 #include <linux/dma-debug.h>
 #include <linux/dma-attrs.h>
+#include <linux/hwiotrace.h>
 #include <asm-generic/dma-coherent.h>
 
 static inline dma_addr_t dma_map_single_attrs(struct device *dev, void *ptr,
@@ -262,6 +263,15 @@ static inline void *dma_alloc_attrs(struct device *dev, size_t size,
 
 	cpu_addr = ops->alloc(dev, size, dma_handle, flag, attrs);
 	debug_dma_alloc_coherent(dev, size, *dma_handle, cpu_addr);
+
+#if defined(CONFIG_HWIOTRACE) && defined(CONFIG_KHWIO_CONSISTENT_DMA)
+	register_hwiotrace_probe(
+		dev_name(dev),
+		cpu_addr,
+		size,
+		DMA_CONSIST);
+#endif
+
 	return cpu_addr;
 }
 
@@ -271,6 +281,12 @@ static inline void dma_free_attrs(struct device *dev, size_t size,
 {
 	const struct dma_map_ops *ops = get_dma_ops(dev);
 
+#if defined(CONFIG_HWIOTRACE) && defined(CONFIG_KHWIO_CONSISTENT_DMA)
+	unregister_hwiotrace_probe(
+		dev_name(dev),
+		cpu_addr);
+#endif
+
 	BUG_ON(!ops);
 	WARN_ON(irqs_disabled());
 
